opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java

```diff
@@ -210,13 +220,101 @@ private ExprValue constructTimestamp(String value) {
     }
   }
 
-  private ExprValue parseTimestamp(Content value) {
+  // returns java.time.format.Parsed
+  private TemporalAccessor parseTimestampString(String value, OpenSearchDateType dt) {
+    for (var formatter : dt.getRegularFormatters()) {
+      try {
+        return formatter.parse(value);
+      } catch (Exception ignored) {
+        // nothing to do, try another format
+      }
+    }
+    for (var formatter : dt.getNamedFormatters()) {
+      try {
+        return formatter.parse(value);
+      } catch (Exception ignored) {
+        // nothing to do, try another format
+      }
+    }
+    return null;
+  }
+
+  private ExprValue formatReturn(ExprType formatType, ExprTimestampValue unformatted) {
+    if (formatType.equals(ExprCoreType.DATE)) {
+      return new ExprDateValue(unformatted.dateValue());
+    }
+    if (formatType.equals(ExprCoreType.DATETIME)) {
+      return new ExprDatetimeValue(unformatted.datetimeValue());
+    }
+    if (formatType.equals(ExprCoreType.TIME)) {
+      return new ExprTimeValue(unformatted.timeValue().toString());
+    }
+    return unformatted;
+  }
+
+  private ExprValue parseTimestamp(Content value, ExprType type) {
+    OpenSearchDateType dt;
+    ExprType returnFormat;
+    if (type instanceof OpenSearchDateType) {
+      dt = (OpenSearchDateType) type;
+      returnFormat = dt.getExprType();
+    } else {
+      dt = OpenSearchDateType.of();
+      returnFormat = ((OpenSearchDataType) type).getExprType();
+    }
+
     if (value.isNumber()) {
-      return new ExprTimestampValue(Instant.ofEpochMilli(value.longValue()));
+      return formatReturn(
+          returnFormat,
+          new ExprTimestampValue(Instant.ofEpochMilli(value.longValue())));
     } else if (value.isString()) {
-      return constructTimestamp(value.stringValue());
+      TemporalAccessor parsed = parseTimestampString(value.stringValue(),dt);
+      if (parsed == null) { // failed to parse or no formats given
+        return formatReturn(
+            returnFormat,
+            (ExprTimestampValue)constructTimestamp(value.stringValue()));
+      }
+      // Try Timestamp
+      try {
+        return formatReturn(returnFormat, new ExprTimestampValue(Instant.from(parsed)));
+      } catch (DateTimeException ignored) {
+        // nothing to do, try another type
+      }
+      //Try Datetime
+      try {
+        return formatReturn(
+            returnFormat,
+            new ExprTimestampValue(
+                new ExprDatetimeValue(LocalDateTime.from(parsed)).timestampValue()));
+      } catch (DateTimeException ignored) {
+        // nothing to do, try another type
+      }
+      //Try Time
+      try {
+        return formatReturn(
+            returnFormat,
+            new ExprTimestampValue(
+                new ExprTimeValue(LocalTime.from(parsed))
+                    .timestampValue(new FunctionProperties(Instant.EPOCH, ZoneOffset.UTC))));
```

Comment:
It works as `LocalDate.now()` I suppose.
We need a way to use here already created for this query instance of `FunctionProperties`.